src/kbmod/regionsearch/utilities.py

```diff
@@ -0,0 +1,316 @@
+        else:
+            for dirname in ["tmp", "data"]:
+                # sections must be grouped together
+                if not self.data_loaded:
+                    data_loaded_count = 0
+                    for section in sections:
+                        filename = self.__filename(dirname, basename, section[0], suffix)
+                        if os.path.exists(filename) and self.read_table(filename, format, list(section[1])):
+                            data_loaded_count += 1
+                        else:
+                            # early out since this is not a complete set of data
+                            break
+                    if data_loaded_count == len(sections):
+                        self.data_loaded = True
+                        # early out since this is a complete set of data
+                        break
+
+            # if data are not loaded, generate them
+            if not self.data_loaded:
+                self.generate_and_save(basename, suffix, format, sections)
+
+        if not self.data_loaded:
+            raise Exception(f"Could not read or generate data")
+
+    def generate_and_save(self, basename, suffix, format, sections):
+        """
+        Generate the data and save them in the tmp directory.
+
+        Parameters
+        ----------
+        basename : str
+        """
+        if self._generate_data():
+            if os.path.isdir("tmp"):
+                for section in sections:
+                    filename = self.__filename("tmp", basename, section[0], suffix)
+                    self.write_table(filename, format, section[1])
+            self.data_loaded = True
+            self.data_generated = True
+
+    def __filename(self, dirname, basename, sectionname, suffix):
+        """
+        Return the full filename.
+
+        Parameters
+        ----------
+        dirname : str
+            The directory name. This is "tmp" or "data".
+        basename : str
+            The base name of the file to read or write.
+        sectionname : str
+            The section name.
+        suffix : str
+            The suffix of the file to read or write. Thisis the last part of the file name.
+
+        Returns
+        -------
+        str
+            The full filename.
+        """
+        return f"{dirname}/{basename}-{self.seed}-{self.clustercnt}-{self.samplespercluster}-{self.version}-{sectionname}{suffix}"
+
+    def read_table(self, filename: str, format: str, colnames: typing.List[str]):
+        """
+        Read the table from the file and validate the version.
+
+        Parameters
+        ----------
+        filename : str
+            The full filename.
+        format : str
+            The format of the file to read or write.
+        colnames : typing.List[str]
+            The list of column names to read from the file. The file may contain other columns but it must have these columns.
+        Returns
```

Comment:
Add a space before the "Returns" section.

Reply:
done